### ISSUE TITLE
关于损失函数的一些问题

### ISSUE THREAD

**Issue by Jaqen00** [czifan/DeepSurv.pytorch#3]
我刚开始学习Pytorch，有点没太看懂您的损失函数，方便解释一下嘛。
我把您的损失函数和DeepSurv的损失函数分别计算了一个固定值，但结果却不一样。
后来我自己又写了一个损失函数，结果貌似能对的上。
我是新手，对自己的代码没信心==！，想请您帮忙看一下我的损失函数是否有问题，如果可以也可以分步解释一下您的损失函数嘛，哈哈。主要没弄懂那个mask参数的意义。
代码如下：
Loss function of DeepSurv:
```
import torch
import theano.tensor as T
import numpy
import theano
def _negative_log_likelihood(E,risk):
    # risk = self.risk(deterministic)  # h_hat
    hazard_ratio = T.exp(risk)  # e^h_hat
    log_risk = T.log(T.extra_ops.cumsum(hazard_ratio))  # log(cumulative sum of e^h_hat)
    uncensored_likelihood = risk.T - log_risk  # e^h_hat - log(cumulative sum of e^h_hat)
    # return uncensored_likelihood
    censored_likelihood = uncensored_likelihood * E  # (e^h_hat - log(cumulative sum of e^h_hat))*E
    num_observed_events = T.sum(E)  # total num of observed event
    neg_likelihood = -T.sum(
        censored_likelihood) / num_observed_events  # -(sum of (e^h_hat - log(cumulative sum of e^h_hat))*E)/total num of observed event
    return neg_likelihood
E = T.vector('E')
risk = T.vector('risk')
z = _negative_log_likelihood(E,risk)
f = theano.function([E,risk],z,on_unused_input='ignore')
print(f(numpy.array([0,1,0,1],dtype='float32'),numpy.array([1.2291e-01,1.4153e+00,-7.9396e-02,-8.2728e-02],dtype='float32')))
#print:1.1422242
```

Loss function in your code:
```
def forward(risk_pred, y, e,):
    mask = torch.ones(y.shape[0], y.shape[0])
    mask[(y.T - y) > 0] = 0
    log_loss = torch.exp(risk_pred) * mask
    log_loss = torch.sum(log_loss, dim=0) / torch.sum(mask, dim=0)
    log_loss = torch.log(log_loss).reshape(-1, 1)
    neg_log_loss = -torch.sum((risk_pred - log_loss) * e) / torch.sum(e)
    print(neg_log_loss)
risk_pred = torch.tensor([1.2291e-01,1.4153e+00,-7.9396e-02,-8.2728e-02],dtype=torch.float32)
y = torch.tensor([4,3,2,1],dtype=torch.float32)
e = torch.tensor([0,1,0,1],dtype=torch.float32)
forward(risk_pred, y, e,)#print:tensor(-1.2891)
```
Loss function I wrote:
```
def loss(events,durations,risk_pred):
    idx = durations.sort(descending=True)[1]
    events = events[idx]
    risk_pred = risk_pred[idx]
    events = events.float()
    events = events.view(-1)
    risk_pred = risk_pred.view(-1)
    uncensored_likelihood = risk_pred - risk_pred.exp().cumsum(0).log()
    censored_likelihood = uncensored_likelihood * events
    num_observed_events = events.sum()
    neg_likelihood = -censored_likelihood.sum()/num_observed_events
    print(neg_likelihood)
risk_pred = torch.tensor([1.2291e-01,1.4153e+00,-7.9396e-02,-8.2728e-02],dtype=torch.float32)
y = torch.tensor([4,3,2,1],dtype=torch.float32)
e = torch.tensor([0,1,0,1],dtype=torch.float32)
loss(e,y,risk_pred)#print:tensor(1.1422)
```

**Comment by czifan** [comment]
你好！谢谢您的提问。
1) 首先，我认为我和DeepSurv的计算结果不同主要是因为我在每一步都做了平均（即除以torch.sum(XXX)），这样做的原因是我在实验过程中发现DeepSurv的代码不好收敛，我总结的原因是cumsum导致loss的数值范围较大的原因。但有可能是我代码写得不好或者实验不合理造成的，如果您能够发现问题所在，请您能够告诉我，让我能够进行修正和学习。
2) 另外，代码的组织思路不同也导致您觉得代码很不一样的原因。我是以[1]中公式(3)为参考的，逐行代码解释如下：
def forward(risk_pred, y, e,):
    mask = torch.ones(y.shape[0], y.shape[0]) # 例如y的维度为(N,1)，那么mask为(N,N)
    mask[(y.T - y) > 0] = 0 # 判断y中的一个位置是否比其他位置的值小（因为这里预测的是风险概率，因此风险概率越大则生存时间越短），例如，y=array([[1], [2], [3]])，那么mask=array([[1., 0., 0.], [1., 1., 0.], [1., 1., 1.]])
    log_loss = torch.exp(risk_pred) * mask # 计算每一个位置的exp，而后乘上mask，即把那些不属于R(T_i)的位置变成0
    log_loss = torch.sum(log_loss, dim=0) / torch.sum(mask, dim=0) # 以行为单位将每一个位置相加，这里因为把不属于R(T_i)的位置都变成0，那么相加就只是对属于R(T_i)的位置相加
    log_loss = torch.log(log_loss).reshape(-1, 1) # 根据公式中，相加后取log
    neg_log_loss = -torch.sum((risk_pred - log_loss) * e) / torch.sum(e) # 根据公式进行计算
    print(neg_log_loss)
[1] DeepSurv: Personalized Treatment Recommender System Using A Cox Proportional Hazards Deep Neural Network
希望这个回答对您有所帮助，如果有错误的地方麻烦您能够指出来！
